rotkehlchen/globaldb/handler.py

```diff
@@ -1550,77 +1554,75 @@ def hard_reset_assets_list(
-                write_cursor.switch_foreign_keys('ON')
-
-                with user_db.user_write() as user_db_cursor:
-                    user_db_cursor.switch_foreign_keys('OFF')
-                    user_db_cursor.execute('DELETE FROM assets;')
-                    # Get ids for assets to insert them in the user db
-                    write_cursor.execute('SELECT identifier from assets')
-                    ids = write_cursor.fetchall()
-                    ids_proccesed = ', '.join([f'("{id[0]}")' for id in ids])
-                    user_db_cursor.execute(f'INSERT INTO assets(identifier) VALUES {ids_proccesed};')  # noqa: E501
-                    user_db_cursor.switch_foreign_keys('ON')
-
-            with user_db.conn.read_ctx() as cursor:
-                # Update the owned assets table
-                user_db.update_owned_assets_in_globaldb(cursor)
-
-        except sqlite3.Error as e:
-            with GlobalDBHandler().conn.read_ctx() as read_cursor:
-                read_cursor.execute(detach_database)
-            log.error(f'Failed to restore assets in globaldb due to {str(e)}')
-            return False, 'Failed to restore assets. Read logs to get more information.'
-
         with GlobalDBHandler().conn.read_ctx() as read_cursor:
-            read_cursor.execute(detach_database)
+            # First check that the operation can be made. If the difference is not the
+            # empty set the operation is dangerous and the user should be notified.
+            with user_db.user_write() as user_db_cursor:
+                diff_ids = GlobalDBHandler().get_user_added_assets(
+                    cursor=read_cursor,
+                    user_db_write_cursor=user_db_cursor,
+                    user_db=user_db,
+                    only_owned=True,
+                )
+            if len(diff_ids) != 0 and not force:
+                msg = 'There are assets that can not be deleted. Check logs for more details.'  # noqa: E501
+                return False, msg
+
+        with GlobalDBHandler().packaged_db_lock:
+            try:
+                with GlobalDBHandler().conn.read_ctx() as read_cursor:
+                    read_cursor.execute(f'ATTACH DATABASE "{builtin_database}" AS clean_db;')
+                    # Check that versions match
+                    query = read_cursor.execute('SELECT value from clean_db.settings WHERE name="version";')  # noqa: E501
+                    version = query.fetchone()
+                    if version is None or int(version[0]) != globaldb_get_setting_value(read_cursor, 'version', GLOBAL_DB_VERSION):  # noqa: E501
+                        msg = (
+                            'Failed to restore assets. Global database is not '
+                            'updated to the latest version'
+                        )
+                        return False, msg
+
+                with GlobalDBHandler().conn.write_ctx() as write_cursor:
```

Comment:
We had this double cursor I believe because the write cursor blocks and we were doing some reads. Since you added another with statement maybe we could use a single

```python
with (
    GlobalDBHandler().packaged_db_lock
    GlobalDBHandler().conn.write_ctx() as write_cursor
):
```

to remove indentation from this part of the code as I don't believe is such a critical thing to execute two more sql queries with a write cursor. You would be able to remove the cursor from the `finally` statement

Reply:
If we move `write_cursor` out of the try-except scope, then no revert will happen in case of an error. So I would not do that